INVALID_MEDIA_RESPONSE時の動画DLリカバリ（画質維持・音声同等フォールバック）

## 概要

動画DLが `ERR::INVALID_MEDIA_RESPONSE` で失敗した際、**画質を維持したまま**同一品質の別URL（`backup_urls` + playurl refetch）でリカバリを試み、全候補で失敗した場合のみ現行通りエラー停止する仕組みの導入。音声（`download_audio_with_fallback`）と同等のリカバリ試行を動画にも展開するが、**別品質ストリームへの切り替え（画質ダウン）は行わない**点が音声とは異なる。

実装は別セッションで行う（本 issue に設計・方針を完備して引き継ぐ）。

## ⚠️ ブロッカー

- **#482 の完了が必須**。本 issue は #482 Phase 2 で導入される playurl refetch インフラ（`refetch_dash_urls` / `refetch_durl_url`、retry_download クロージャの attempt 伝達）に依存する。#482 マージ前に実装・検証は行わないこと。

## 背景・現象

### 現象
- 動画DL（`download_video`）で `ERR::INVALID_MEDIA_RESPONSE` が発生し即座に失敗する
- 動確ログ（2026-07-21）:
  ```
  [WARN] download_url: segment 0 download failed (cdn_idx=0)
  [WARN] retry_download: attempt 1/3 failed: 1 segment(s) failed
  [ERROR] download_url: downloaded size too small: 0 bytes (min 1024 bytes) - likely error response
  [WARN] retry_download: non-retryable: ERR::INVALID_MEDIA_RESPONSE
  [ERROR] [FE] downloadVideo: failed ... ERR::INVALID_MEDIA_RESPONSE
  ```
- 現状、`ERR::INVALID_MEDIA_RESPONSE` は retry_download（`bilibili.rs:2018`）で `ERR::` プレフィックスにより **non-retryable 判定で即中断** され、playurl refetch の機会すら得られない
- DL開始〜失敗が10秒。署名期限切れ（120分）のパターンではなく、CDNエッジ固有の問題（issue #467 の HTTP 200 エラーボディ系）の可能性が高い

### 動機
- 音声側は `download_audio_with_fallback`（`bilibili.rs:1198-`）でリカバリ済み。動画側だけ未実装の設計不整合
- #482 は `ERR::NETWORK`（segment failed）対策が主眼で、`INVALID_MEDIA_RESPONSE` のリカバリは明示的にスコープ外。本 issue は #482 では解消されない事象を扱う

## 調査結果（原因）

### INVALID_MEDIA_RESPONSE の発生
- `downloads.rs:603-611` で `final_downloaded < MIN_MEDIA_BYTES`（= 0バイト < 1024バイト）のとき `ERR::INVALID_MEDIA_RESPONSE` を返却
- issue #467（HTTP 200 + エラーボディ）対策の専用サイズ下限チェック。`final_downloaded != total` チェックより前に実行

### 即中断の仕組み
- retry_download（`bilibili.rs:2017-2024`）は `msg.contains("ERR::")` で **すべての `ERR::` 系を non-retryable** として即 `return Err`
- INVALID_MEDIA_RESPONSE は `ERR::` プレフィックスのため、attempt 2/3 に進まず即座に失敗

### 音声 fallback の実態（参照実装）
- `download_audio_with_fallback`（`bilibili.rs:1198-`）は primary 失敗時、`all_audio_streams`（**別品質の音声ストリーム**）を順に試す（`bilibili.rs:1264`）
- `primary_quality_id` → `stream.id` への品質遷移をログ出力（`bilibili.rs:1210-1213`, `1297-1300`）。**音声は品質変動を許容**
- コメント（`bilibili.rs:1250-1254`）:「tied to a specific URL/CDN edge ... a different stream URL may still succeed」

### 動画で画質を維持して使える別URL候補
- `backup_urls`（別CDNホスト群）: download_url 内で CDN ローテーション（cdn_idx）**実装済み**。今回の事象は backup_urls を含め全滅
- playurl refetch による新署名URL: **#482 Phase 2 で導入予定**
- 動画の `dash.video` は品質ごとに1エントリが通常で、別品質ストリームへの切り替えは画質ダウンを伴う → ユーザー要件（画質維持）と両立しないため採用しない

## 修正方針（確定・ユーザー承認済み）

1. retry_download の non-retryable 判定を洗練: `INVALID_MEDIA_RESPONSE` は **リトライ/refetch 対象に許可**、それ以外の `ERR::`（CANCEL / DISK_FULL / FILE_EXISTS 等の真の業務エラー）は従来通り non-retryable
   - **ホワイトリスト方式**（INVALID_MEDIA_RESPONSE のみ許可）で安全側に倒す
2. #482 の playurl refetch インフラを動画側 INVALID_MEDIA_RESPONSE リカバリに拡張
3. **画質は維持**（別品質ストリームへの切り替えは行わない）
4. 全候補（backup_urls + refetch×N）で失敗したら現行通り `ERR::INVALID_MEDIA_RESPONSE` でエラー停止
5. 真の業務エラーは引き続き即中断（リソース無駄遣い防止）

## 実装計画（暫定・#482 完了後に詳細化）

> #482 の refetch インフラ設計に依存するため、#482 マージ後に本セクションを確定する。

- retry_download の `ERR::` 判定（`bilibili.rs:2018`）をホワイトリスト化: `INVALID_MEDIA_RESPONSE` をリトライ許可、他は non-retryable 維持
- 動画の retry_download クロージャ（#482 で `move |attempt|` + refetch 化済み）を、INVALID_MEDIA_RESPONSE 発生時にも refetch が回るよう調整
- リトライ上限・バックオフは #482 の retry_download 設計に準拠（MAX_ATTEMPTS=3 前提で再検討）
- コミット分割方針: #482 マージ後に決定

## 変更ファイルリスト（暫定）

| ファイル | 変更 |
|---|---|
| `src-tauri/src/handlers/bilibili.rs` | retry_download の non-retryable 判定のホワイトリスト化、動画クロージャの INVALID_MEDIA_RESPONSE リカバリ拡張 |
| `src-tauri/src/utils/downloads.rs` | （必要に応じて）INVALID_MEDIA_RESPONSE 発生時の追加コンテキスト |

※ #482 完了後に最終化

## リスク・懸念

1. **#482 への依存**: refetch インフラが #482 で想定通り完成することが前提。#482 の設計変更が本件に影響する
2. **リカバリ可能性**: INVALID_MEDIA_RESPONSE が CDNエッジ固有の問題（音声コメントの根拠）であれば同品質の別URL/refetch でリカバリ期待。ただし bilibili CDN 全体の問題・レートリミットの場合はリカバリ不可。実地検証が必須
3. **リトライ遅延**: INVALID_MEDIA_RESPONSE をリトライ対象にすると、全滅時のユーザー待ち時間が延びる（現状は即停止）。上限設計とログで進捗を可視化する
4. **non-retryable 判定の安全性**: `ERR::` を安易に除外すると真の業務エラーまでリトライしかねない。ホワイトリスト方式（INVALID_MEDIA_RESPONSE のみ）で安全側に倒す
5. **FE進捗**: リトライ中の進捗バー挙動は #482 の `emit_retrying` 設計に準拠（進捗バー0に戻らない）

## テスト計画

### Rust
- INVALID_MEDIA_RESPONSE 発生時に refetch が呼ばれること
- refetch でも失敗時、最終的に `ERR::INVALID_MEDIA_RESPONSE` で停止すること
- 他の `ERR::`（CANCEL / DISK_FULL / FILE_EXISTS）は引き続き即中断すること（ホワイトリスト検証）
- 画質が維持されること（別品質ストリームに切り替わっていないこと）

### 手動検証
- メディアエラー発生時のリカバリログ（refetch attempt、別URL試行）確認
- 全候補失敗時のエラー停止確認
- `npm run typecheck` / `npm run lint` / `cargo build` / `npm run test` 全て緑

## 参照

- **ブロッカー**: #482（DLネットワークエラー改善とERR::NETWORK UI表示修正）
- 音声参照実装: `download_audio_with_fallback`（`bilibili.rs:1198-`）
- 即中断箇所: retry_download（`bilibili.rs:2017-2024`）
- エラー発生箇所: `downloads.rs:603-611`
- bilibili API 仕様: `references/bilibili-API-collect/docs/video/videostream_url.md`（URL有効期限120分、L102）

## 備考

- 実装は **#482 完了後に別セッションで実施**（本 issue の設計に従う）
- 動確ログ（2026-07-21）で発見。#482 では解消されない事象
- worktree-start 使用時、args に実装内容を書かないこと（forked 実行が設計を無視する既知の問題: MEMORY.md feedback-worktree-start-args）

--- a/src-tauri/src/handlers/bilibili.rs
+++ b/src-tauri/src/handlers/bilibili.rs
@@ -2004,6 +2004,46 @@ mod tests {
         assert_eq!(video.title, "a_b");
         assert_eq!(video.parts[0].sanitized_part.as_deref(), Some("a_b"));
     }
+
+    /// Whitelist for retryable ERR:: codes (issue #484): only
+    /// ERR::INVALID_MEDIA_RESPONSE may retry; other business errors stay
+    /// non-retryable.
+    #[test]
+    fn test_is_retryable_err_code() {
+        assert!(is_retryable_err_code("ERR::INVALID_MEDIA_RESPONSE"));
+        // Wrapped / suffixed variants still match (contains-based, matching
+        // how download_url errors may carry appended detail).
+        assert!(is_retryable_err_code(
+            "prefix ERR::INVALID_MEDIA_RESPONSE detail"
+        ));
+
+        assert!(!is_retryable_err_code("ERR::CANCELLED"));
+        assert!(!is_retryable_err_code("ERR::DISK_FULL"));
+        assert!(!is_retryable_err_code("ERR::FILE_EXISTS"));
+        assert!(!is_retryable_err_code("ERR::MERGE_FAILED"));
+        assert!(!is_retryable_err_code("ERR::NETWORK::2 segment(s) failed"));
+        assert!(!is_retryable_err_code("connection reset by peer"));
+        assert!(!is_retryable_err_code(""));
+    }
+
+    /// Exhaustion keeps whitelisted ERR:: codes verbatim (so callers and
+    /// the UI classify them correctly) and wraps transient messages as
+    /// ERR::NETWORK (issue #484).
+    #[test]
+    fn test_exhausted_retry_error() {
+        assert_eq!(
+            exhausted_retry_error("ERR::INVALID_MEDIA_RESPONSE".to_string()),
+            "ERR::INVALID_MEDIA_RESPONSE"
+        );
+        assert_eq!(
+            exhausted_retry_error("2 segment(s) failed".to_string()),
+            "ERR::NETWORK::2 segment(s) failed"
+        );
+        assert_eq!(
+            exhausted_retry_error("final size mismatch: 10 vs 20".to_string()),
+            "ERR::NETWORK::final size mismatch: 10 vs 20"
+        );
+    }
 }
 
 /// Spawns an async task to save download history.
@@ -3259,7 +3299,11 @@ fn ensure_free_space(target_path: &Path, needed_bytes: u64) -> Result<(), String
 ///
 /// - `ERR::` prefix: Business logic errors (e.g. `ERR::DISK_FULL`,
 ///   `ERR::CANCELLED`, `ERR::FILE_EXISTS`) are passed through immediately
-///   without retry.
+///   without retry. Exception: whitelisted codes
+///   (`ERR::INVALID_MEDIA_RESPONSE`, issue #484) are CDN-edge-scoped error
+///   bodies tied to a specific signed URL, so they are retried — the next
+///   attempt re-fetches a fresh signed playurl URL (see the closures at
+///   each call site).
 /// - All other errors: Treated as transient network failures and retried.
 ///   `download_url` only produces non-`ERR::` errors for network-related
 ///   causes (request failures, connection resets, timeouts, segment issues),
@@ -3271,7 +3315,9 @@ fn ensure_free_space(target_path: &Path, needed_bytes: u64) -> Result<(), String
 ///
 /// - Maximum attempts: 3
 /// - Backoff strategy: Linear (500ms, 1000ms, 1500ms)
-/// - Final failure is wrapped as `ERR::NETWORK::{original_message}`
+/// - Final failure is wrapped as `ERR::NETWORK::{original_message}`;
+///   whitelisted `ERR::` codes keep their own code so callers and the UI
+///   can classify them (issue #484)
 ///
 /// # Retry State Notification
 ///
@@ -3296,8 +3342,13 @@ fn ensure_free_space(target_path: &Path, needed_bytes: u64) -> Result<(), String
 /// # Errors
 ///
 /// Returns errors in the following cases:
-/// - All retry attempts failed (wrapped as `ERR::NETWORK::*`)
-/// - Error contains `ERR::` prefix (passed through unchanged)
+/// - All retry attempts failed (wrapped as `ERR::NETWORK::*`; whitelisted
+///   `ERR::` codes keep their own code)
+/// - Error contains a non-whitelisted `ERR::` prefix (passed through
+///   unchanged)
+///
+/// See also: `is_retryable_err_code` / `exhausted_retry_error` for the
+/// whitelist and exhaustion-wrapping decisions (issue #484).
 async fn retry_download<F, Fut>(
     app: &AppHandle,
     download_id: &str,
@@ -3337,8 +3388,11 @@ where
             Err(e) => {
                 let msg = e.to_string();
 
-                // ERR:: prefix = business logic error, never retry
-                if msg.contains("ERR::") {
+                // ERR:: prefix = business logic error, never retry.
+                // Exception: whitelisted codes (issue #484) are CDN-edge
+                // error bodies — fall through so the next attempt re-fetches
+                // a fresh signed playurl URL before re-downloading.
+                if msg.contains("ERR::") && !is_retryable_err_code(&msg) {
                     log::warn!("[BE] retry_download: non-retryable: {msg}");
                     if attempt > 1 {
                         emit_retrying(false);
@@ -3346,14 +3400,15 @@ where
                     return Err(msg);
                 }
 
-                // Non-ERR:: errors from download_url are network-related.
-                // Retry unconditionally; final attempt wraps as ERR::NETWORK.
+                // Transient network errors and whitelisted ERR:: codes retry;
+                // the final attempt wraps as ERR::NETWORK, except whitelisted
+                // codes which keep their own semantic code.
                 if attempt >= MAX_ATTEMPTS {
                     log::error!("[BE] retry_download: exhausted {MAX_ATTEMPTS} attempts: {msg}");
                     if attempt > 1 {
                         emit_retrying(false);
                     }
-                    return Err(format!("ERR::NETWORK::{msg}"));
+                    return Err(exhausted_retry_error(msg));
                 }
 
                 log::warn!("[BE] retry_download: attempt {attempt}/{MAX_ATTEMPTS} failed: {msg}");
@@ -3363,6 +3418,26 @@ where
     }
 
     unreachable!()
+}
+
+/// True for `ERR::` codes that are CDN-edge-scoped (error body tied to a
+/// specific signed URL) and worth retrying with a re-fetched playurl URL.
+/// All other `ERR::` codes are true business errors and stay non-retryable.
+/// Whitelist intentionally minimal (issue #484).
+fn is_retryable_err_code(msg: &str) -> bool {
+    msg.contains("ERR::INVALID_MEDIA_RESPONSE")
+}
+
+/// Final error after exhausting all retry attempts. Only whitelisted
+/// `ERR::` codes reach exhaustion (non-whitelisted ones return early);
+/// they keep their semantic code while transient network messages are
+/// wrapped as `ERR::NETWORK` (issue #484).
+fn exhausted_retry_error(msg: String) -> String {
+    if msg.contains("ERR::") {
+        msg
+    } else {
+        format!("ERR::NETWORK::{msg}")
+    }
 }
 
 /// Selects a stream URL from the quality list.

